snuba/cli/query_fetcher.py

```diff
@@ -0,0 +1,210 @@
+
+import click
+import structlog
+
+from snuba.clickhouse.native import ClickhousePool
+from snuba.clusters.cluster import ClickhouseClientSettings
+from snuba.environment import setup_logging, setup_sentry
+from snuba.utils.gcs import GCSUploader
+
+logger = structlog.get_logger().bind(module=__name__)
+
+QueryToRerun = namedtuple("QueryToRerun", ["query_str", "query_id"])
+
+
+class QueriesFileSaver:
+    def __init__(self, gcs_bucket: Optional[str] = None) -> None:
+        self.uploader = None
+        if gcs_bucket:
+            try:
+                self.uploader = GCSUploader(gcs_bucket)
+            except Exception as e:
+                logger.info(f"Couldn't set up gcs for bucket {gcs_bucket}: {str(e)}")
+
+    def _format_filename(self, table: str, date: datetime) -> str:
+        # Example: queries_2024_01_16_1_errors_local - first hour
+        #          queries_2024_01_16_2_errors_local- second hour
+        day = datetime.strftime(date, "%Y_%m_%d")
+        hour = date.hour
+        return f"queries_{day}_{hour}_{table}"
+
+    def _format_blob_name(self, table: str, date: datetime) -> str:
+        # Example: queries/2024_01_16/1/errors_local - first hour
+        #          queries/2024_01_16/2/errors_local- second hour
+        day = datetime.strftime(date, "%Y_%m_%d")
+        hour = date.hour
+        return f"queries/{day}/{hour}/{table}"
+
+    def _full_path(self, filename: str) -> str:
+        return f"/tmp/{filename}.csv"
+
+    def _save_to_csv(self, filename: str, results: Sequence[QueryToRerun]) -> None:
+        with open(self._full_path(filename), mode="w") as file:
+            writer = csv.writer(file)
+            for row in results:
+                writer.writerow(row)
+
+        logger.info(f"File {self._full_path(filename)} saved")
+
+    def _save_to_gcs(self, filename: str, blob_name: str) -> None:
+        if self.uploader:
+            self.uploader.upload_file(self._full_path(filename), blob_name)
+
+    def save(self, table: str, date: datetime, results: Sequence[QueryToRerun]) -> None:
+        """
+        First save the results to local csv file,
+        then upload the file to gcs bucket.
+        """
+        filename = self._format_filename(table, date)
+        self._save_to_csv(filename, results)
+
+        blob_name = self._format_blob_name(table, date)
+        self._save_to_gcs(filename, blob_name)
+
+
+def get_querylog_query(
+    databases: list[str], tables: list[str], start: datetime, end: datetime
+) -> str:
+    start_time = datetime.strftime(start, "%Y-%m-%d %H:%M:%S")
+    end_time = datetime.strftime(end, "%Y-%m-%d %H:%M:%S")
+
+    return rf"""
+    SELECT
+        query_id,
+        query
+    FROM system.query_log
+    WHERE (query_kind = 'Select')
+    AND (type = 'QueryFinish')
+    AND (databases = {databases})
+    AND (tables = {tables})
+    AND (query_start_time >= toDateTime('{start_time}'))
+    AND (query_start_time <= toDateTime('{end_time}'))
+    LIMIT 100
```

Comment:
Any reason for the limit? What would happen in cases where there are more than 100 queries? I do not see any sort of pagination happening.

Reply:
this was just supposed to be for testing, not sure if we want to have limits and handle pagination now

Reply:
If we want to get all queries to get compared within the time range we should fix this. Either remove limit. But then I do not know if clickhouse would apply some limit. Or add in some sort of pagination with limit and offset. Probably easier to do. That way you ensure that you are able to verify all queries.

Reply:
Removing the limit for now - might have to come back to this though